python/lib/communication/dmod/communication/maas_request/ngen/ngen_exec_request_body.py

```diff
@@ -35,7 +58,37 @@ class Config:
 
     def dict(self, **kwargs) -> dict:
         # if exclude is set, ignore this _get_exclude_fields()
-        only_if_set = ("catchments", "partition_cfg_data_id", "forcings_data_id")
+        only_if_set = ("catchments", "partition_cfg_data_id", "forcings_data_id", "partial_realization_config",
+                       "t_route_config_data_id")
         if kwargs.get("exclude", False) is False:
             kwargs["exclude"] = {f for f in only_if_set if not self.__getattribute__(f)}
         return super().dict(**kwargs)
+
+    @property
+    def composite_config_source_ids(self) -> List[str]:
+        """
+        A list of the data ids for any datasets that are sources of data for a generated composite config dataset.
```

Comment:
This is specifically to get dataset ids for datasets that are sources when we are creating a new composite config dataset, if there are any. Those will be put into the data requirement describing what kind of composite dataset that the job needs, which will eventually lead to certain things being done in the deriver util.

Reply:
Yeah, the part where im likely being overly literal is if a composite config dataset id is provided in the request body and no other datasets are included, this will return the empty list. But it would be confusing, at least to me, that the datasources from the provided composite dataset are not returned. I think I am just being overly pedantic about this method's name. Would you consider something instead? The only thing I can come up with at the moment is `non_composite_config_ids`, but I also dont really like that.

Reply:
I don't think there is any way around this.  In general, we can't know the details of an existing composite config dataset, beyond its `data_id`, at the time a request object is created.  

It may also help to think of this not as describing the dataset, but describing a data requirement for the job being requested that some dataset must fulfill.  If the requirement for config data doesn't include restrictions on source datasets, then that's perfectly valid.

Reply:
> I don't think there is any way around this. In general, we can't know the details of an existing composite config dataset, beyond its `data_id`, at the time a request object is created.

Yeah, I agree that this we _need_ this, I am just not so fond of the name. However, if it is this difficult to come up with a different name, the current is probably our best option.

Reply:
Would `source_dataset_ids` be a better name?  Or perhaps `build_from_datasets`?

Reply:
I think the current name is appropriate if it is not allowed to provide a `realization_config_data_id`, `bmi_config_data_id`, `t_route_config_data_id`, _and_ an existing `composite_config_data_id`. So, should we enforce the invariant that you cannot create a `NGENRequestBody` instance using if `composite_config_data_id` and a `realization_config_data_id`, `bmi_config_data_id`, or `t_route_config_data_id` is provided?

Reply:
I started to add a validator for this invariant, but after thinking more, I'm not sure it applies.

The purpose of these types (one at least) is to generally describe the data and datasets that a job needs. The request and request body objects can't really know whether a dataset exists with some _data_id_; they only describe what's necessary.  It isn't strictly invalid to describe the necessary dataset as being a composite of certain others _and_ having a specific name.  

Perhaps a user wants or needs to differentiate what is required for a job by specifying a particular unique composite config dataset name, to force something new to be created (e.g., so that freshly generated BMI init configs are used).

Reply:
> We _can_ still add a validator for this; I'm just not certain it's appropriate. And given there is need for an overhaul of some of this anyway, it may be just as wise to wait until then.

That's fair. We can address this at a later time.